crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -2820,6 +2745,121 @@ enum ForkedDependencies {
+    /// The package(s) that provoked at least one additional fork.
+    diverging_packages: Vec<PackageName>,
+}
+
+impl Forks {
+    /// Add the given dependency to this fork.
+    fn add(&mut self, dep: PubGrubDependency) {
+        use crate::marker::{is_definitively_empty_set, is_disjoint};
+
+        // A root can never be a dependency of another package,
+        // and a `Python` pubgrub package is never returned by
+        // `get_dependencies`. So a pubgrub package always has a
+        // name in this context.
+        let name = dep.package.name().expect("dependency always has a name");
+        let mut markers = dep
+            .package
+            .marker()
+            .cloned()
+            .unwrap_or(MarkerTree::And(vec![]));
+        if is_definitively_empty_set(&markers) {
+            // A marker that we know never matches any environment
+            // can be skipped entirely because its dependency will
+            // never be included in any circumstance.
+            return;
+        } else if self.forks.is_empty() {
+            // If we don't have any forks yet, we just seed it with
+            // two forks from `markers` and `not(markers)`.
+            if !markers.is_universal() {
+                self.diverging_packages.push(name.clone());
+                self.forks.push(Fork {
+                    dependencies: vec![],
+                    markers: markers.negate(),
+                });
+            }
+            self.forks.push(Fork {
+                dependencies: vec![dep],
+                markers,
+            });
+            return;
+        } else if markers.is_universal() {
+            // If our markers match every marker environment,
+            // then the dependency should be added to every fork.
+            for fork in &mut self.forks {
+                fork.dependencies.push(dep.clone());
+            }
+            return;
+        }
+        // Otherwise, we *may* need to add a new fork...
+
+        let mut new = vec![];
+        let mut found_overlap = false;
+        for mut fork in std::mem::take(&mut self.forks) {
+            if is_disjoint(&fork.markers, &markers) {
+                new.push(fork);
+                continue;
+            }
+            found_overlap = true;
+
+            let not_markers = markers.negate();
+            let mut new_markers = markers.clone();
+            new_markers.and(fork.markers.negate());
+            if !is_definitively_empty_set(&not_markers) && !is_disjoint(&fork.markers, &not_markers)
+            {
+                new.push(Fork {
+                    markers: {
+                        let mut fork_markers = fork.markers.clone();
+                        fork_markers.and(not_markers);
+                        fork_markers
+                    },
+                    dependencies: fork.dependencies.clone(),
+                });
+            }
+            if !is_definitively_empty_set(&markers) && !is_disjoint(&fork.markers, &markers) {
```

Comment:
Haven't we guaranteed with the `is_disjoint(&fork.markers, &markers)` check that this can never be empty?